client/client.go

```diff
@@ -1396,169 +1478,175 @@ func (c *Client) deriveToken(alloc *structs.Allocation, taskNames []string, vcli
-				break
+			if len(servers) > 0 {
+				break DISCOLOOP
 			}
 		}
-		if len(nomadServerServices) == 0 {
-			if len(mErr.Errors) > 0 {
-				return mErr.ErrorOrNil()
-			}
-
-			return fmt.Errorf("no Nomad Servers advertising service %q in Consul datacenters: %q", nomadServerServiceName, dcs)
+	}
+	if len(servers) == 0 {
+		if len(mErr.Errors) > 0 {
+			return mErr.ErrorOrNil()
 		}
+		return fmt.Errorf("no Nomad Servers advertising service %q in Consul datacenters: %+q", serviceName, dcs)
+	}
 
-		// Log the servers we are adding
-		c.logger.Printf("[INFO] client.consul: bootstrap adding following Servers: %q", nomadServerServices)
+	c.logger.Printf("[INFO] client.consul: discovered following Servers: %s", servers)
+	c.servers.set(servers)
 
-		c.heartbeatLock.Lock()
-		if atomic.LoadInt32(&c.lastHeartbeatFromQuorum) == 1 && now.Before(c.consulPullHeartbeatDeadline) {
-			c.heartbeatLock.Unlock()
-			// Common, healthy path
-			if err := c.rpcProxy.SetBackupServers(nomadServerServices); err != nil {
-				return fmt.Errorf("client.consul: unable to set backup servers: %v", err)
-			}
-		} else {
-			c.heartbeatLock.Unlock()
-			// If this Client is talking with a Server that
-			// doesn't have a leader, and we have exceeded the
-			// consulPullHeartbeatDeadline, change the call from
-			// SetBackupServers() to calling AddPrimaryServer()
-			// in order to allow the Clients to randomly begin
-			// considering all known Nomad servers and
-			// eventually, hopefully, find their way to a Nomad
-			// Server that has quorum (assuming Consul has a
-			// server list that is in the majority).
-			for _, s := range nomadServerServices {
-				c.rpcProxy.AddPrimaryServer(s)
+	// Notify waiting rpc calls. Wait briefly in case initial rpc
+	// just failed but the calling goroutine isn't selecting on
+	// discovered yet.
+	const dur = 50 * time.Millisecond
+	timeout := time.NewTimer(dur)
+	for {
+		select {
+		case c.discovered <- struct{}{}:
+			if !timeout.Stop() {
+				<-timeout.C
 			}
+			timeout.Reset(dur)
+		case <-timeout.C:
+			return nil
 		}
-
-		return nil
 	}
-	if c.config.ConsulConfig.ClientAutoJoin {
-		c.consulSyncer.AddPeriodicHandler("Nomad Client Fallback Server Handler", bootstrapFn)
-	}
-
-	consulServicesReaperFn := func() error {
-		const estInitialExecutorDomains = 8
-
-		// Create the domains to keep and add the server and client
-		domains := make([]consul.ServiceDomain, 2, estInitialExecutorDomains)
-		domains[0] = consul.ServerDomain
-		domains[1] = consul.ClientDomain
+}
 
-		for allocID, ar := range c.getAllocRunners() {
-			ar.taskStatusLock.RLock()
-			taskStates := copyTaskStates(ar.taskStates)
-			ar.taskStatusLock.RUnlock()
-			for taskName, taskState := range taskStates {
-				// Only keep running tasks
-				if taskState.State == structs.TaskStateRunning {
-					d := consul.NewExecutorDomain(allocID, taskName)
-					domains = append(domains, d)
-				}
+func (c *Client) consulReaper() {
```

Comment:
Fixed.

Good call; there was no need for the ticker to be on the struct at all!